prometheus/vec.go

```diff
@@ -231,19 +243,115 @@ func (m *MetricVec) hashLabels(labels Labels) (uint64, error) {
 		if !ok {
 			return 0, fmt.Errorf("label name %q missing in label map", label)
 		}
-		h = hashAdd(h, val)
+		h = m.hashAdd(h, val)
 	}
 	return h, nil
 }
 
-func (m *MetricVec) getOrCreateMetric(hash uint64, labelValues ...string) Metric {
-	metric, ok := m.children[hash]
+// getOrCreateMetric retrieves the metric by hash and label value or creates it
+// and returns the new one.
+//
+// This function takes a lock, so avoid
+func (m *MetricVec) getOrCreateMetric(hash uint64, values interface{}) Metric {
+	m.mtx.RLock()
+	metric, ok := m.getMetric(hash, values)
+	m.mtx.RUnlock()
+	if ok {
+		return metric
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	metric, ok = m.getMetric(hash, values)
 	if !ok {
 		// Copy labelValues. Otherwise, they would be allocated even if we don't go
 		// down this code path.
-		copiedLabelValues := append(make([]string, 0, len(labelValues)), labelValues...)
-		metric = m.newMetric(copiedLabelValues...)
-		m.children[hash] = metric
+		labelValues := m.copyLabelValues(values)
+		metric = m.newMetric(labelValues...)
+		m.children[hash] = append(m.children[hash], metricLabelValue{values: labelValues, metric: metric})
 	}
 	return metric
 }
+
+// getMetric while handling possible collisions in the hash space. Must be
+// called while holding read mutex.
+func (m *MetricVec) getMetric(h uint64, values interface{}) (Metric, bool) {
+	metrics, ok := m.children[h]
+	if ok {
+		return m.selectMetric(metrics, values)
+	}
+
+	return nil, false
+}
+
+func (m *MetricVec) selectMetric(metrics []metricLabelValue, values interface{}) (Metric, bool) {
+	i := m.findMetric(metrics, values)
+
+	if i < len(metrics) {
+		return metrics[i].metric, true
+	}
+
+	return nil, false
+}
+
+// findMetric returns the index of the matching metric or len(metrics) if not
+// found.
+func (m *MetricVec) findMetric(metrics []metricLabelValue, values interface{}) int {
+	for i, metric := range metrics {
+		if m.matchLabels(metric.values, values) {
+			return i
+		}
+	}
+
+	return len(metrics)
+}
+
+func (m *MetricVec) matchLabels(values []string, labels interface{}) bool {
```

Comment:
Most importantly, the `interface{}` passed throughout the call stack is always the same, so especially in lack of any type hints, it would help to name the function parameter the same in all functions.